### ISSUE TITLE
Add quarter tiling window snapping

### ISSUE THREAD

**Issue by Jacalz** [fyne-io/fynedesk#118]
Arguably the best thing about windows is that it was the first to introduce quarter tiling window snapping. Most Linux window manager do this too, however some like Mutter in Gnome only do so with left and right in dual tiling. If we are to implement it, we ought to do it good from the start :)

## The solution would look something like this:

Dragging of window:
- Dragging to right edge of screen should tile the window to the right half of the screen.
- Dragging to left edge of screen should tile the window to the left half of the screen.
- Dragging to the top edge of screen should make it maximized.
- Dragging to any screen corner should tile it to 25% of screen in that corner.
- Window should go back to the size it had before when dragging it away again.

Moving using keyboard (using the super key, aka your windows key):
- `Super + Right` should tile the window to the right half of the screen.
- `Super + Left` should tile the window to the left half of the screen.
- `Super + Left` followed by `Super + Up` should quarter tile to top left 25% of screen (same for up, down and left, right key-presses in the correct 25% corner spot).
- `Super + Down` should go out of of the snapped tiling.
- `Super + Down + Down` should minimize to the icon dock.

All manual dragging should show a highlight of the screen.

To sum it up, we should follow what Windows are doing :)

**Comment by andydotxyz** [comment]
Nice plan.
Not entirely certain about the keyboard shortcuts, but the concept is solid :)

If the dragging requires a highlight overlay then it may be a while as I think that requires a compositor to do well (though up for other suggestions that we could take on!)

**Comment by Jacalz** [comment]
The keyboard shortcuts are great. They are the same as to what Windows 7, 8 and 10 uses along with what basically every Linux window manager uses if they support window snapping :)

Some really good implementations of this would be either Windows or, on the Linux side, Mate and KDE Plasma. Those are the ones that I know support quarter tiling on top of my head and the ones that I think do a good job of how window snapping should behave :+1:

**Comment by andydotxyz** [comment]
Sure thing - sorry I didn't mean to question /having/ keyboard shortcuts. Just wasn't certain if the Windows ones specifically would be used. Figured we maybe should see what is common and/or set up what we wanted to use.
Our other shortcuts are all Alt-based which is why I had pause.

**Comment by Jacalz** [comment]
No worries. I understood what you meant. Though, I do think it is best to stick with what most Linux and Windows users would be accustomed to 🙂

**Comment by mattboll** [comment]
If I may add a shortcut : 
`Super + up` should maximize the window

**Comment by Jacalz** [comment]
Absolutely. I don't know how I could miss that one 😅

**Comment by COD3HUNT3R** [comment]
This might be related. In gnome with mutter I can drag in the window while fullscreen and it automatically exits fullscreen and resizes. But can not do that in fynedesk. Window doesn't move at all.

**Comment by Jacalz** [comment]
> This might be related. In gnome with mutter I can drag in the window while fullscreen and it automatically exits fullscreen and resizes. But can not do that in fynedesk. Window doesn't move at all.

That is already mentioned in the issue description.

**Comment by andydotxyz** [comment]
> This might be related. In gnome with mutter I can drag in the window while fullscreen and it automatically exits fullscreen and resizes. But can not do that in fynedesk. Window doesn't move at all.

You mean maximised right? Fullscreen has no border...

**Comment by COD3HUNT3R** [comment]
> You mean maximised right? Fullscreen has no border...

yes
